File: 01-Numerical_Errors/5-multiplicative_errors.cpp
/* Exploration of multiplicative errors using Bessel functions */
#include <iostream>
#include <cmath>

typedef double number;

number myBessel(number l, number x);

int main(int argc, char **argv)
{
  std::cout.precision(14);
  std::cout.setf(std::ios::scientific);
  number x = 0.5;

  int L=25;
  std::cout << "  n  " << '\t'
            << "my Bessel(n," << x << ")" << '\t'
            << "  j(n," << x << ")  " << '\t'
            << "  | myJ - j |   " << '\n';

  for(number n = 23; n >= 0; n--)
  {
  std::cout << n << '\t' << myBessel(n,x)  << '\t'  <<  std::sph_bessel(n,x) << '\t' << std::abs(myBessel(n,x) - std::sph_bessel(n,x)) << '\n';
  }


}

number myBessel(number l, number x)
{
  unsigned L = 25;
  number JL = std::sph_bessel(L,x);
  number JL1 = std::sph_bessel(L+1,x);
  number J;

  for(int ll = L; ll >= l+1; ll--)
  {
    J = (2*(ll) + 1)*JL/x - JL1;
    JL1 = JL;
    JL = J;
  }
  return J;

}

File: 01-Numerical_Errors/5-multiplicative_errors.py
"""
Multiplicative errors
"""
import numpy as np
import matplotlib.pyplot as plt
from scipy import special


def myBessel(l,x):
    L = 25
    JL = special.spherical_jn(L,x)
    JL1 = special.spherical_jn(L+1,x)
    
    
    for i in np.arange(L,l,-1):
        myJ = (2*i + 1)*JL/x - JL1
        JL1 = JL
        JL = myJ
    return myJ


def main():
    x = 1
    l = 24
    
    BesselArray = np.zeros(l+1)
    MyArray = np.zeros(l+1)
    D = np.zeros(l+1)
    L = np.arange(l,-1,-1)
    
    print(' l ', '   myJ(l,x)    ', '     J(l,x)  ', '   |muJ - J|    ')
    for n in L:
        BesselArray[n] = special.spherical_jn(n,x)
        MyArray[n] = myBessel(n, x)
        D[n] = abs(myBessel(n,x) - special.spherical_jn(n,x) )
        print(n,  
              MyArray[n], 
              BesselArray[n],
              D[n]
              )
        
    L.sort()
    
    plt.xlabel('$n$')
    plt.ylabel('$|j_n{rec} (^x) - j_n^{built}(x)|$')
    plt.title('Multiplicative Errors with Bessel functions')
    
    plt.loglog(L,D,'-*')
    plt.savefig("bessel_py.pdf")
    
    
main()

